modules/core/src/main/java/org/apache/ignite/internal/processors/service/ServiceInfo.java

```diff
@@ -0,0 +1,148 @@
+import org.apache.ignite.services.ServiceDescriptor;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Service's information container.
+ */
+public class ServiceInfo implements ServiceDescriptor {
+    /** */
+    private static final long serialVersionUID = 0L;
+
+    /** Node ID. */
+    private final UUID nodeId;
+
+    /** Service id. */
+    private final IgniteUuid srvcId;
+
+    /** Service configuration. */
+    private final ServiceConfiguration cfg;
+
+    /** Topology snapshot. */
+    @GridToStringInclude
+    private volatile Map<UUID, Integer> top = Collections.emptyMap();
+
+    /**
+     * @param nodeId Initiating node id.
+     * @param srvcId Service id.
+     * @param cfg Service configuration.
+     */
+    public ServiceInfo(@NotNull UUID nodeId, @NotNull IgniteUuid srvcId, @NotNull ServiceConfiguration cfg) {
+        this.nodeId = nodeId;
+        this.srvcId = srvcId;
+        this.cfg = cfg;
+    }
+
+    /**
+     * Sets service's new topology snapshot.
+     *
+     * @param top Topology snapshot.
+     */
+    public void topologySnapshot(@NotNull Map<UUID, Integer> top) {
+        this.top = top;
+    }
+
+    /**
+     * Returns service's configuration.
+     *
+     * @return Service configuration.
+     */
+    public ServiceConfiguration configuration() {
+        return cfg;
+    }
+
+    /**
+     * Rerurns services id.
+     *
+     * @return Service id.
+     */
+    public IgniteUuid serviceId() {
+        return srvcId;
+    }
+
+    /** {@inheritDoc} */
+    @Override public String name() {
+        return cfg.getName();
+    }
+
+    /** {@inheritDoc} */
+    @SuppressWarnings("unchecked")
+    @Override public Class<? extends Service> serviceClass() {
+        if (cfg instanceof LazyServiceConfiguration) {
+            String clsName = ((LazyServiceConfiguration)cfg).serviceClassName();
+
+            try {
+                return (Class<? extends Service>)Class.forName(clsName);
+            }
+            catch (ClassNotFoundException e) {
+                throw new IgniteException("Failed to find service class: " + clsName, e);
+            }
+        }
+        else
+            return cfg.getService().getClass();
+    }
+
+    /** {@inheritDoc} */
+    @Override public int totalCount() {
+        return cfg.getTotalCount();
+    }
+
+    /** {@inheritDoc} */
+    @Override public int maxPerNodeCount() {
+        return cfg.getMaxPerNodeCount();
+    }
+
+    /** {@inheritDoc} */
+    @Nullable @Override public String cacheName() {
+        return cfg.getCacheName();
+    }
+
+    /** {@inheritDoc} */
+    @SuppressWarnings("unchecked")
+    @Nullable @Override public <K> K affinityKey() {
+        return (K)cfg.getAffinityKey();
+    }
+
+    /** {@inheritDoc} */
+    @Override public UUID originNodeId() {
+        return nodeId;
+    }
+
+    /** {@inheritDoc} */
+    @Override public Map<UUID, Integer> topologySnapshot() {
+        return top;
```

Comment:
Do we want to share a reference to this map? 
Should we share a reference to read only view of it?

Reply:
Wrapped by `Collections#unmodifiableMap`